### PR TITLE
Add RegEx checking for individual websites

### PR DESCRIPTION
**Pull request by andrewda**
This will check the desired username against the required pattern
for each website to ensure the username is possible.

Closes https://github.com/manu-chroma/username-availability-checker/issues/11

### CODE REVIEW

File: username_api.py [hunk 1]
```diff
@@ -16,15 +17,37 @@ def check_username(website, username):
...
+			return {'status': code, 'url': url}
+
+		else:
+			return {'status': r.get(url).status_code, 'url': url}
+	else:
+		return {'status': 400, 'url': url}
+
+def check_format(website, username):
+	pattern = {
+	'pinterest'  :r'^[a-zA-Z0-9_]{3,15}$',
```
Comment (jayvdb) [inline username_api.py]: bad indent level.  And IMO this is data, and should be in a JSON file or python module.

Eventually it should be submitted to https://github.com/manu-chroma/username-availability-checker/issues/11#issuecomment-352158545 , and we use their definitions which is now https://github.com/manu-chroma/username-availability-checker/issues/25

File: username_api.py [hunk 2]
```diff
@@ -16,15 +17,37 @@ def check_username(website, username):
...
+				else 404
 
-		return {'status': code, 'url': url}
-	
+			return {'status': code, 'url': url}
+
+		else:
+			return {'status': r.get(url).status_code, 'url': url}
+	else:
+		return {'status': 400, 'url': url}
```
Comment (jayvdb) [inline username_api.py]: this is not a good use of `400`, as it conflicts with actual `400`, which is about the http request , not the application logic.

The client made a correctly framed request, which was able to be correctly understood, and the app can respond sanely to inform the client that their request is invalid due to rules which the client doesnt know about.

File: username_api.py (hunk 1 above)
Comment (andrewda) [reply username_api.py]: I kept indentation similar to the indentation on the existing lines https://github.com/manu-chroma/username-availability-checker/blob/master/username_api.py#L12-L17 – should fix this one and keep the other one how it is?

File: username_api.py (hunk 2 above)
Comment (andrewda) [reply username_api.py]: Agreed, makes sense.

File: websites.py [hunk 3]
```diff
@@ -0,0 +1,10 @@
+patterns = {
+	'pinterest': r'^[a-zA-Z0-9_]{3,15}$',
+	'gitlab': r'^[a-zA-Z0-9_\-.]+$',
+	'github': r'^[a-zA-Z0-9-]{1,39}$',
+	'tumblr': r'^[a-zA-Z0-9-]{1,32}$',
+	'behance': r'^[a-zA-Z0-9_\-]{1,30}$',
+	'instagram': r'^[a-zA-Z0-9_.]+$',
+	'twitter': r'^[a-zA-Z0-9_]{5,15}$',
+	'soundcloud': r'^[a-zA-Z0-9_\-]+$'
+}
```
Comment (jayvdb) [inline websites.py]: EOL at EOF

File: username_api.py [hunk 4]
```diff
@@ -16,15 +19,40 @@ def check_username(website, username):
 	'behance'   :'https://{}.net/{}'.format(website, username)
 	}.get(website, 'https://{}.com/{}'.format(website, username)) # default
 
-	if website in ['pinterest', 'gitlab']:
-		res  = r.get(url)
-		code = 200 if bytes(username, encoding='utf-8') in res.content \
-			else 404
+	possible = check_format(website, username)
+
+	if possible:
```
Comment (jayvdb) [inline username_api.py]: you can avoid doing a re-indentation by doing

File: username_api.py [hunk 5]
```diff
@@ -16,15 +19,40 @@ def check_username(website, username):
...
+			'possible': possible,
+		}
+	else:
+		return {
+			'status': r.get(url).status_code,
+			'url': url,
+			'possible': possible,
+		}
+
+def check_format(website, username):
```
Comment (manu-chroma) [inline username_api.py]: can you add a comment here describing what this function does?

**Comment by manu-chroma** [comment]
Changes look great. Just address the above comment and rebase. :+1:

**Comment by manu-chroma** [comment]
LGTM. @jayvdb do you have any other comments? 

also, pls rebase

File: websites.py [hunk 6]
```diff
@@ -0,0 +1,10 @@
+patterns = {
+	'pinterest': r'^[a-zA-Z0-9_]{3,15}$',
```
Comment (jayvdb) [inline websites.py]: I think it is more useful to blow this regex into parts in the data here.  have min and max lengths as separate fields, and store a "valid characters" list containing `a-zA-Z0-9_`.

Then at start build the compiled regexes using `^[{valid_characters}]{min,max}$'`.

`min` of 1 is implicit.

Error if max isnt provided.

File: websites.py [hunk 7]
```diff
@@ -0,0 +1,10 @@
...
+	'gitlab': r'^[a-zA-Z0-9_\-.]+$',
```
Comment (jayvdb) [inline websites.py]: ahem
https://gitlab.com/gitlab-org/gitlab-ce/issues/40116#note_47256905

File: websites.py [hunk 8]
```diff
@@ -0,0 +1,10 @@
...
+	'github': r'^[a-zA-Z0-9-]{1,39}$',
+	'tumblr': r'^[a-zA-Z0-9-]{1,32}$',
+	'behance': r'^[a-zA-Z0-9_\-]{1,30}$',
+	'instagram': r'^[a-zA-Z0-9_.]+$',
```
Comment (jayvdb) [inline websites.py]: there will be a limit.  Something will break somewhere.  If you cant find any sane limit, use 999 and include a link in a comment showing that the max length isnt known

File: websites.py [hunk 9]
```diff
@@ -0,0 +1,10 @@
...
+	'twitter': r'^[a-zA-Z0-9_]{5,15}$',
+	'soundcloud': r'^[a-zA-Z0-9_\-]+$'
```
Comment (jayvdb) [inline websites.py]: per instagram

File: username_api.py (hunk 5 above)
Comment (jayvdb) [reply username_api.py]: (ftr, this was done, but github review comment is pinned to a visible line.)

File: websites.py [hunk 10]
```diff
@@ -0,0 +1,10 @@
+patterns = {
```
[the remark this answers is not in the document]
Comment (jayvdb) [reply websites.py]: If you want to keep calling it website patterns, also include the website URLs pattern overrides currently at https://github.com/manu-chroma/username-availability-checker/pull/28/files#diff-86cae0060e15d8121c1f9029f69f7517L13

File: websites.py (hunk 9 above)
Comment (manu-chroma) [reply websites.py]: Instagram doesn't let you input more than ``30`` chars in the username field.

**Comment by manu-chroma** [comment]
Thanks @andrewda

**Pull request merged**